File: java/sort/MergeSort.java
package sort;

import java.util.Arrays;

public class MergeSort {
    public static int[] mergeSort(int[] arr) {
        if (arr.length <= 1) {
            return arr;
        }

        int mid = arr.length / 2;
        int[] leftHalf = Arrays.copyOfRange(arr, 0, mid);
        int[] rightHalf = Arrays.copyOfRange(arr, mid, arr.length);

        int[] sortedLeft = mergeSort(leftHalf);
        int[] sortedRight = mergeSort(rightHalf);

        return merge(sortedLeft, sortedRight);
    }

    public static int[] merge(int[] left, int[] right) {
        int[] result = new int[left.length + right.length];
        int i = 0, j = 0, k = 0;

        while (i < left.length && j < right.length) {
            if (left[i] < right[j]) {
                result[k++] = left[i++];
            } else {
                result[k++] = right[j++];
            }
        }

        while (i < left.length) {
            result[k++] = left[i++];
        }

        while (j < right.length) {
            result[k++] = right[j++];
        }

        return result;
    }


    public static int[] mergeWithoutRecursion(int[] left, int[] right) {
        int[] result = new int[left.length + right.length];
        int i = 0, j = 0, k = 0;

        while (i < left.length && j < right.length) {
            if (left[i] < right[j]) {
                result[k++] = left[i++];
            } else {
                result[k++] = right[j++];
            }
        }

        while (i < left.length) {
            result[k++] = left[i++];
        }

        while (j < right.length) {
            result[k++] = right[j++];
        }

        return result;
    }

    public static int[] mergeSortWithoutRecursion(int[] arr) {
        int step = 1;
        int length = arr.length;

        while (step < length) {
            for (int i = 0; i < length; i += 2 * step) {
                int mid = Math.min(i + step, length);
                int end = Math.min(i + 2 * step, length);

                int[] left = Arrays.copyOfRange(arr, i, mid);
                int[] right = Arrays.copyOfRange(arr, mid, end);

                int[] merged = merge(left, right);

                System.arraycopy(merged, 0, arr, i, merged.length);
            }
            step *= 2;
        }

        return arr;
    }
}

File: python/sort/MergeSort.py
class MergeSort:
    @staticmethod
    def mergeSort(arr):
        if len(arr) <= 1:
            return arr

        mid = len(arr) // 2
        leftHalf = arr[:mid]
        rightHalf = arr[mid:]

        sortedLeft = MergeSort.mergeSort(leftHalf)
        sortedRight = MergeSort.mergeSort(rightHalf)

        return MergeSort.merge(sortedLeft, sortedRight)

    @staticmethod
    def merge(left, right):
        result = []
        i = j = 0

        while i < len(left) and j < len(right):
            if left[i] < right[j]:
                result.append(left[i])
                i += 1
            else:
                result.append(right[j])
                j += 1

        result.extend(left[i:])
        result.extend(right[j:])

        return result
    

    @staticmethod
    def mergeSortWithoutRecursion(arr):
        step = 1  
        length = len(arr)
        
        while step < length:
            for i in range(0, length, 2 * step):
                left = arr[i:i + step]
                right = arr[i + step:i + 2 * step]
                
                merged = MergeSort.merge(left, right)
                
                for j, val in enumerate(merged):
                    arr[i + j] = val
                    
            step *= 2 
            
        return arr